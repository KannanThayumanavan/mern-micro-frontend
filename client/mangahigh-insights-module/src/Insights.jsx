import React, { useEffect, useState } from 'mangahigh-libs/react';
import Summary from './components/Summary';
import AttemptsDataTable from './components/AttemptsDataTable';
import axios from 'axios';

const Insights = ({}) => {
  const [attempts, setAttempts] = useState(null);

  useEffect(() => {
    axios
      .get("http://localhost:5000/attempts")
      .then((res) => {
        const sortResponseBasedOnDate = res.data
          .sort((obj1, obj2) => new Date(obj2.date_and_time) - new Date(obj1.date_and_time));          
        setAttempts(sortResponseBasedOnDate);
      })
  }, []);

  return(
    <div className="container-fluid p-3">
      <div className="row text-center">
        <div className="col-md-12">
        {!attempts && (
          <div className="spinner-border text-primary" role="status">
            <span className="visually-hidden">Loading...</span>
          </div>  
        )}
        {attempts && (
          <div className="p-3 border bg-light">
            <Summary attempts={attempts} />
            <AttemptsDataTable attempts={attempts} />
          </div>
        )}
        </div>
      </div>
    </div>  
  );
};

export default Insights;