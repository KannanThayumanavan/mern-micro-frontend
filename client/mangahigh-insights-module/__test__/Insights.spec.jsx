import React from 'react';
import { shallow, mount } from 'enzyme';
import Insights from '../src/Insights';
import axios from "axios";
import { act } from "react-dom/test-utils";

jest.mock('axios');

jest.mock('mangahigh-libs/react', () => jest.requireActual('react'),
	{ virtual: true }
);

jest.mock('mangahigh-components/DisplayCard', () => () => 'DisplayCard'			
, { virtual: true });

jest.mock('mangahigh-components/Table', () => () => 'Table'			
, { virtual: true });

const response = {
	data: [
		{
			"_id":"6161b01c74f86b0917b418c8",
			"date_and_time":"2021-10-09T15:07:07.763Z",
			"score":3,
			"total_questions":3
		},
		{
			"_id":"6161b01c74f86b0917b418b6",
			"date_and_time":"2021-10-09T15:37:07.763Z",
			"score":1,
			"total_questions":3
		}
	],
}

describe('Test Insights Component', () => {
	let renderedModule;
	afterEach(() => {
    jest.clearAllMocks();
  });

	it('should render loading state correctly', () => {
		renderedModule = shallow(<Insights />);
		expect(renderedModule.find('.spinner-border').exists()).toBeTruthy();
    expect(renderedModule).toMatchSnapshot();
	});

	it('should render Insights component with succesfull fetch call', async () => {
		await act(async () => {
      await axios.get.mockImplementationOnce(() => Promise.resolve(response));  
      renderedModule = mount(<Insights />);    
    });		
		renderedModule.update();
		await expect(axios.get).toHaveBeenCalledTimes(1);
		expect(renderedModule.find('Summary').exists()).toBeTruthy();
		expect(renderedModule.find('AttemptsDataTable').exists()).toBeTruthy();		
		expect(renderedModule).toMatchSnapshot();
	});
});