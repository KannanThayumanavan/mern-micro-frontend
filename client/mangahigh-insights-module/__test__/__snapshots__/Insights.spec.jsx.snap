// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Insights Component should render Insights component with succesfull fetch call 1`] = `
<Insights>
  <div
    className="container-fluid p-3"
  >
    <div
      className="row text-center"
    >
      <div
        className="col-md-12"
      >
        <div
          className="p-3 border bg-light"
        >
          <Summary
            attempts={
              Array [
                Object {
                  "_id": "6161b01c74f86b0917b418b6",
                  "date_and_time": "2021-10-09T15:37:07.763Z",
                  "score": 1,
                  "total_questions": 3,
                },
                Object {
                  "_id": "6161b01c74f86b0917b418c8",
                  "date_and_time": "2021-10-09T15:07:07.763Z",
                  "score": 3,
                  "total_questions": 3,
                },
              ]
            }
          >
            <div
              className="row"
            >
              <div
                className="col-md-6"
              >
                <Component
                  highlight="2"
                  summary="Number of times played"
                >
                  DisplayCard
                </Component>
              </div>
              <div
                className="col-md-6"
              >
                <Component
                  highlight="2.0"
                  summary="Average score"
                >
                  DisplayCard
                </Component>
              </div>
            </div>
          </Summary>
          <AttemptsDataTable
            attempts={
              Array [
                Object {
                  "_id": "6161b01c74f86b0917b418b6",
                  "date_and_time": "2021-10-09T15:37:07.763Z",
                  "score": 1,
                  "total_questions": 3,
                },
                Object {
                  "_id": "6161b01c74f86b0917b418c8",
                  "date_and_time": "2021-10-09T15:07:07.763Z",
                  "score": 3,
                  "total_questions": 3,
                },
              ]
            }
          >
            <div
              className="row"
            >
              <div
                className="col-md-12"
              >
                <h4
                  className="text-left"
                >
                  Historical Results 
                  <small>
                    (latest on top)
                  </small>
                </h4>
                <Component
                  tableData={
                    Object {
                      "header": Array [
                        "#",
                        "Date and time of play",
                        "Score",
                        "Total Questions",
                      ],
                      "rows": Array [
                        Array [
                          "1",
                          "Sat, 09 Oct 2021 15:37:07 GMT",
                          "1",
                          "3",
                        ],
                        Array [
                          "2",
                          "Sat, 09 Oct 2021 15:07:07 GMT",
                          "3",
                          "3",
                        ],
                      ],
                    }
                  }
                >
                  Table
                </Component>
              </div>
            </div>
          </AttemptsDataTable>
        </div>
      </div>
    </div>
  </div>
</Insights>
`;

exports[`Test Insights Component should render loading state correctly 1`] = `
<div
  className="container-fluid p-3"
>
  <div
    className="row text-center"
  >
    <div
      className="col-md-12"
    >
      <div
        className="spinner-border text-primary"
        role="status"
      >
        <span
          className="visually-hidden"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;
